python/desc/twinkles/sprinkler.py

```diff
@@ -185,255 +184,317 @@ def __init__(self, catsim_cat, visit_mjd, specFileMap, sed_path,
-                    #Convert orientation angle to west of north from east of north by *-1.0 and convert to radians
-                    row[self.defs_dict['galaxyBulge_positionAngle']] = newlens['PHIE']*(-1.0)*np.pi/180.0
-
-                    if self.logging_is_sprinkled:
-                        row[self.defs_dict['galaxyAgn_is_sprinkled']] = 1
-                        row[self.defs_dict['galaxyBulge_is_sprinkled']] = 1
-                        row[self.defs_dict['galaxyDisk_is_sprinkled']] = 1
-
-                    #Replace original entry with new entry
-                    updated_catalog[rowNum] = row
+                rng = np.random.RandomState(galtileid % (2^32 -1))
+                pick_value = rng.uniform()
+
+                if len(candidates) == 0 or pick_value>self.density_param:
+                    # If there aren't any lensed sources at this redshift from
+                    # OM10 move on the next object
+                    continue
+
+                # Randomly choose one the lens systems
+                # (can decide with or without replacement)
+                # Sort first to make sure the same choice is made every time
+
+                candidates = candidates[np.argsort(candidates['twinklesId'])]
+                newlens = rng.choice(candidates)
             else:
-                if self.cached_sprinkling is True:
-                    if galtileid in self.sne_cache['galtileid'].values:
-                        use_system = self.sne_cache.query('galtileid == %i' % galtileid)['twinkles_system'].values
-                        use_df = self.sne_catalog.query('twinkles_sysno == %i' % use_system)
-                        self.used_systems.append(use_system)
-                    else:
-                        continue
+                twinkles_sys_cache = self.agn_cache.query('galtileid == %i' % galtileid)['twinkles_system'].values[0]
+                newlens = self.lenscat[np.where(self.lenscat['twinklesId'] == twinkles_sys_cache)[0]][0]
+
+            #varString = json.loads(row[self.defs_dict['galaxyAgn_varParamStr']])
+            # varString[self.defs_dict['pars']]['t0_mjd'] = 59300.0
+            #row[self.defs_dict['galaxyAgn_varParamStr']] = json.dumps(varString)
+
+            # Append the lens galaxy
+            # For each image, append the lens images
+            default_lensrow = None
+            if newlens['NIMG'] > 0:
+                default_lensrow = row.copy()
+                default_lensrow[self.defs_dict['galaxyDisk_majorAxis']] = 0.0
+                default_lensrow[self.defs_dict['galaxyDisk_minorAxis']] = 0.0
+                default_lensrow[self.defs_dict['galaxyDisk_positionAngle']] = 0.0
+                default_lensrow[self.defs_dict['galaxyDisk_internalAv']] = 0.0
+                default_lensrow[self.defs_dict['galaxyDisk_magNorm']] = 999. #np.nan To be fixed post run1.1
```

Comment:
Should we remove these notes to ourselves about handling the magNorms? These are working so that in the final instance catalogs we only have the disk and bulge entries from `bulge/disk_agn_host` right?

Reply:
So: the concern is that we want to make sure that the galaxies that are in the catalog pre-sprinkler are totally removed...

I will add a test to the InstanceCatalog validation script to just verify that none of the values of `galaxy_id` specified in the `cosmoDC2_v1.1.4_agn_cache.csv` appear in the InstanceCatalogs.  Will that address the concern?

Reply:
Yeah, I think that would be good. Thanks.

Reply:
I pushed a change to `validate_ic.py` yesterday.  Verifying that none of the replaced `galaxy_id`s appear isn't quite right, because the lens galaxies keep the old `galaxy_id`.  It's only the multiply-imaged AGN that get new `galaxy_id`s.  So, the new test

- Verifies that no `galaxy_id`s greater than 10**11 exist in the Sersic profile InstanceCatalogs (since this would indicate that the bulge/disk components of the multiply lensed AGN were making it into the InstanceCatalog)

-Verifies that none of the point sources have `magNorm>900` or `magNorm==None`, since that would indicate that CatSim was not correctly handling the placeholder magNorms.

The commit in question is here

https://github.com/LSSTDESC/Twinkles/pull/495/commits/2d6f5c6f4cec43fdb3d0b836d17587826b2a7e54

Do you agree with my reasoning?

Reply:
Yes, this looks right. The host bulge/disk components from Matt's code go in a separate file, right?

Reply:
Yes.  There is a separate InstanceCatalog file for the FITS-based components